### DISCUSSION TITLE
quering the wifi mode (station, softap, both)

### DISCUSSION THREAD

**Discussion by atesin** [JAndrassy/WiFiEspAT#85] [General]
hi... i need to know if my esp is working on softap mode (not exclusive)... my first instinct tell me to use the `status()`command  
- the [Enhanced WiFi API](https://github.com/jandrassy/WiFiEspAT#enhanced-wifi-api) section tells me this lib is used similar to [WiFiNINA library](https://www.arduino.cc/en/Reference/WiFiNINA)
- [WiFi.status() documentation](https://www.arduino.cc/reference/en/libraries/wifinina/wifi.status/) shows many return states, i am interested in `WL_AP_LISTENING` or `WL_AP_CONNECTED`
- looking the source code i noticed `WL_AP_LISTENING` is just used once in `WiFiClass::beginAP()`, while `WL_AP_CONNECTED`isn't even mentioned
- so looking for some `AT+CWMODE_CUR?`/`AT+CWMODE?` command wrapper, the only i saw is in `EspAtDrvClass::reset()` function

so i think i have currently no way to know if my softap is running or not (maybe the same with station mode)

looking the source, i noticed the `status()` function is mainly targeted to station mode... and talking about WiFiNINA api, i dont know what status it would return me if i have both station+softap modes enabled, is ambiguous... or both modes running but no clients connected, or both modes running but one with clients and the other no... maybe it should have 2 statuses, one for station mode and another one for softap mode

**Comment by atesin** [comment]
i posted a [similar question](https://github.com/arduino-libraries/WiFiNINA/issues/221) in WiFiNINA repo site

**Comment by JAndrassy** [comment]
as the library doesn't set SoftAP persistent start, the AP only runs if the sketch started it.
the same applies for WiFiNina.

**Comment by atesin** [comment]
if the esp has commands like AT+CWMODE_DEF and AT+AUTOCONN to start (softap mode / station mode / autoconnect ) ... which do you think is the criteria in WiFiEspAT and WiFiNINA to **NOT** set these modes/connections start at boot?? ...  this way who need are forced to plug an ftdi and send CWMODE_DEF by hand as the libs dont set it

if somebody dont want the mode to be saved just dont use setPersistent()  ...  as currently setPersistent() dont saves all the modes so is not fully functional, in my humble point of view ... could you clarify this criteria a little more? .. thanks

**Comment by JAndrassy** [comment]
SetupPersistentWiFiConnection has
```
  WiFi.setPersistent();
  WiFi.endAP(); // to disable default automatic start of persistent AP at startup
```
endAP(true) or endAP() with setPersistent disable automatic start of SoftAP

**Comment by atesin** [comment]
maybe i didn't explain myself enough, this is my situation, maybe somebody else have a similar one

i am developing a system that works together between arduino and my server (http + php + mariadb), a low cost irrigation system, but also could be a domotics or an automation one

when the farmer receives the device (arduino + esp8266) he must provision his wifi for the system starts to run.... then the wifi access should be saved and started automatically, and the farmer (often not so skilled in technology) should NOT enter wifi credentials everytime the system shuts down (by energy failures, signal lost or whatever)

so my question actually is... how can i make the esp8266 save wifi credentials and CONNECT AUTOMATICALLY when boot using this library... because i don't understand why it wasn't designed to auto-connect, a flaw imho

---
eehhmm ... backing the original question  xD  ... is there some wrapper for AT+CWMODE_CUR? to query which mode the esp is currently running? (station and/or **softap**).. or with another commands for each mode or something? ....

in some conditions i need to know if softap mode is running (with or without clients), my further sketch running flow depends on this

**Comment by JAndrassy** [comment]
https://github.com/khoih-prog/WiFiManager_Generic_Lite
or the ConfigurationAP example

**Comment by atesin** [comment]
> The AVR-family boards (UNO, Nano, Mega, etc.) are not supported

i had seen ConfigurationAP example but it doesn't satisfy these questions:
- with this library, how can i know if SoftAP mode is active or not (with or without clients)?
- why WiFiEspAT and WiFiNINA libraries can't set a SoftAP to start automatically when boot?

**Comment by JAndrassy** [comment]
>     * with this library, how can i know if SoftAP mode is active or not (with or without clients)?

you can't. but you can stop it and then it is not running or start it and then it is running. STA connection can drop so we need a status() function. but we have full control over SoftAP.
 
>     * why WiFiEspAT and WiFiNINA libraries can't set a SoftAP to start automatically when boot?

the STA connection takes time so it is better for WiFiEspAT to let the esp start it as soon as possible and do it asynchronous to setup(), while SoftAP starts almost immediately.

**Comment by atesin** [comment]
hi.... summarizing, we can say ESP has 2 independent "network interfaces", station and softap, and there are commands for each according the scope and osi layer

![esp-inerfaces](https://user-images.githubusercontent.com/8086106/180628189-46c3daed-1d22-48a2-90f3-95989bf6ea1c.png)
_here each physical link is represented by a wave_

below there are the "status" commands i think, excluding layer 5 (session, ssl) and services (dhcp, dns, ntp, etc.)

### layer 1 (wlan parameters)

physical links are negotiated so scope doesn't really apply here, i think the equivalent '/etc/init.d/network status' command which tells which interface is "up" or "down" (bitwise), is `AT+CWMODE_CUR?`, along with these to get parameters:

- station: `AT+CWJAP_CUR?`
- softap: `AT+CWSAP_CUR?`

### local scope

layer 2 (ethernet, mac address)
- station: `AT+CIPSTAMAC_CUR?`
- softap: `AT+CIPAPMAC_CUR?`

layer 3 (ip + routes)
- station: `AT+CIPSTA_CUR?`
- softap: `AT+CIPAP_CUR?`

layer 2+3
- station+softap: `AT+CIFSR`

layer 4 (tcp+udp sockets, only has meaning in remote scope)

### remote scope

layer 2
- station: `AT+CWJAP_CUR?` (`<bssid>` often matches remote ap mac address)
- softap: `AT+CWLIF` (connected station's mac addresses in `<mac>`parameter)

layer 3
- station: couldn't found a way to get remote ap ip address, but should match with `<gateway>` in `AT+CIPSTA_CUR?` (*)
- softap: `AT+CWLIF` (connected station's ip addresses in `<ip addr>`parameter)

(*) i believe for administration access, "routers" have pins connected to its internal switch, so are just like another node in the segment

layer 4
station+softap: `AT+CIPSTATUS`

...

as we can see, a single `status()` function is way insufficient to get all the information that could be gathered, considering just the esp8266 dual "network interfaces"

i understand you tried to make this library compatible with WiFiNINA api, but you know wifinina by itself is not enough to cover all esp8266 needs, you make it clear when added the `accept()`, `startMDNS()` and other functions, so don't be afraid to add more functions and get this lib even better (than wif...) ;) ... for example regarding this discussion, a `statusAP()` function could be added

**Comment by JAndrassy** [comment]
if you make a good pull request, I merge it
